File: java/program/DataController.java
import java.util.Scanner;
import java.util.ArrayList;


public class DataController {

    // wadah data
    private ArrayList<Shirt> container = new ArrayList<>();
    private Scanner sc = new Scanner(System.in);


    public DataController() {

    }

    public DataController(ArrayList<Shirt> src) {
        this.container = src;
    }

    public ArrayList<Shirt> getContainer() {
        return this.container;
    }

    public void setContainer(ArrayList<Shirt> container) {
        this.container = container;
    }

    // method mencari data dari wadah berdasarkan id
    public Shirt searchById(String id) {
        for(Shirt obj : this.container) {
            if(obj.getIdProduct().equals(id)) {
                return obj;
            }
        }
        return null;
    }

    // menambah data ke wadah
    public void createData() {
        String id, name, brand, size, material, gender, color, sleeveType;
        int price;

        // ambil input attribute attribute dari user
        System.out.println("\nPlease give attributes for new data:");
        System.out.print("Product ID  : ");
        id = sc.next();
        sc.nextLine();
        System.out.print("Name        : ");
        name = sc.nextLine();
        System.out.print("Brand       : ");
        brand = sc.nextLine();
        System.out.print("Price       : ");
        price = sc.nextInt();
        sc.nextLine();
        System.out.print("Size        : ");
        size = sc.nextLine();
        System.out.print("Material    : ");
        material = sc.nextLine();
        System.out.print("Gender      : ");
        gender = sc.nextLine();
        System.out.print("Color       : ");
        color = sc.nextLine();
        System.out.print("Sleeve Type :");
        sleeveType = sc.nextLine();

        // instansiasi objek dengan attribut yang sudah diberikan
        Shirt temp = new Shirt(id, name, brand, price, size, material, gender, color, sleeveType);
        // sambungkan objek ke wadah
        this.container.add(temp);

        System.out.println("Data has been successfully created!");
        

    }

    public void modifyData() {
        String targetID;
        // minta id data yang ingin di modifikasi
        System.out.print("\nPlease select data to modify by ID : ");
        targetID = sc.next();

        // cari datanya & cek apakah ditemukan
        Shirt target = this.searchById(targetID);
        if(target == null) {
            System.err.println("Data was not found!");  // jika tidak ditemukan, keluar dari method modifikasi
            return; 
        }

        // jika ditemukan data, lanjut minta attribut attribut baru untuk data yang dipilih
        System.out.println("\nData found. Please enter new data attributes");
        sc.nextLine();

        int price;
        String name, brand, size, material, gender, color, sleeveType;

        System.out.print("Name        : ");
        name = sc.nextLine();
        System.out.print("Brand       : ");
        brand = sc.nextLine();
        System.out.print("Price       : ");
        price = sc.nextInt();
        System.out.print("Size        : ");
        size = sc.nextLine();
        System.out.print("Material    : ");
        material = sc.nextLine();
        System.out.print("Gender      : ");
        gender = sc.nextLine();
        System.out.print("Color       : ");
        color = sc.nextLine();
        System.out.print("Sleeve Type :");
        sleeveType = sc.nextLine();

        // ubah attribut data
        target.setName(name);
        target.setBrand(brand);
        target.setPrice(price);
        target.setSize(size);
        target.setMaterial(material);
        target.setGender(gender);
        target.setColor(color);
        target.setSleeveType(sleeveType);

        System.out.println("Data successfully modified!");

    }

    public void deleteData() {
        // cari data berdasarkan id
        String targetID;
        System.out.print("\nPlease select data to delete by ID : ");
        targetID = sc.next();

        // cari dan cek apakah ditemukan
        Shirt target = this.searchById(targetID);
        if(target == null) {
            System.err.println("Data was not found!");  // jika tidak ditemukan maka keluar dari method
            return;
        }

        // jika ditemukan datanya

        System.out.println("Data has been found.");
        // menampilkan nama data yg ditemukan 
        System.out.println(target.getIdProduct() + ". " + target.getName());
        // konfirmasi penghapusan data
        System.out.println("Confirm to delete? (y/n)");

        char del_confirm = sc.next().charAt(0);
        // cek apakah dikonfirmasi atau tdk
        if(del_confirm == 'y') {
            this.container.remove(target); // hapus data
        } else {
            // penghapusan dibatalkan
            System.out.println("Cancelled data deletion\n");
        }
        
        
    }
    
    public void showAll() {
        Table table = new Table();  // instansiasi sebuah tabel
        int numRows = this.container.size();    // cek berapa baris yang akan dibutuhkan berdasarkan ukuran wadah data
        
        String[][] data = new String[numRows + 1][9];   // buat 2d array string baris data (+1 untuk kepala tabel)* jumlah kolom
        
        // set baris pertama untuk kepala tabel nya
        data[0][0] = "Product ID";
        data[0][1] = "Name";
        data[0][2] = "Brand";
        data[0][3] = "Price";
        data[0][4] = "Size";
        data[0][5] = "Material";
        data[0][6] = "Gender";
        data[0][7] = "Color";
        data[0][8] = "Sleeve Type";




        // ambil value string dari setiap attribut semua data dan disimpan ke 2d string
        for (int i = 0; i < numRows; i++) {
            Shirt Shirt = container.get(i);
            data[i + 1][0] = Shirt.getIdProduct();
            data[i + 1][1] = Shirt.getName();
            data[i + 1][2] = Shirt.getBrand();
            data[i + 1][3] = Integer.toString(Shirt.getPrice());    // konversi integer ke string
            data[i + 1][4] = Shirt.getSize();
            data[i + 1][5] = Shirt.getMaterial();
            data[i + 1][6] = Shirt.getGender();
            data[i + 1][7] = Shirt.getColor();
            data[i + 1][8] = Shirt.getSleeveType();



        }

        // panggil fungsi pembuatan baris tabel
        table.createRows(numRows + 1, 9, data);
        
    }

}

File: cpp/program/DataController.cpp
#pragma once
#include <iostream>
#include <string>
#include <list>
#include "Table.cpp"
#include "Shirt.cpp" // Include the Shirt class header file

using namespace std;

class DataController
{
private:
    list<Shirt> container; // container for all the data

public:

    DataController() {
        // empty constructor, container will be initialized as an empty list
    }

    // getter and setter for container

    list<Shirt>& getContainer() {
        return this->container;
    }

    void setContainer(list<Shirt>& container) {
        this->container = container;
    }

    // searching a specific data using id
    list<Shirt>::iterator searchByID(string id) {

        bool found = false;
        list<Shirt>::iterator it = this->container.begin();    // iterator to find
        list<Shirt>::iterator it_found = it;   // iterator copy to store found data
        while (found != true && it != this->container.end()) {
            if (it->getId() == id) {
                found = true;
                it_found = it;  // store found data location
            }
            it++;
        }

        if (found) {
            return it_found;    // return found data location
        }
        else {
            return this->container.end(); // else return .end() to indicate data wasn't found
        }
    }

    void createData() {

        // temp variables for inputs
        string tempId, tempName, tempBrand, tempSize, tempMaterial, tempGender, tempColor, tempSleeveType;
        double tempPrice;

        cout << "Please give attributes for new data:\n";
        cout << "ID : ";
        cin >> tempId;
        cout << "Name : ";
        cin >> tempName;
        cout << "Brand : ";
        cin >> tempBrand;
        cout << "Price : ";
        cin >> tempPrice;
        cout << "Size : ";
        cin >> tempSize;
        cout << "Material : ";
        cin >> tempMaterial;
        cout << "Gender : ";
        cin >> tempGender;
        cout << "Color : ";
        cin >> tempColor;
        cout << "Sleeve Type : ";
        cin >> tempSleeveType;

        // instantiate a Shirt object with inputted values
        Shirt new_data(tempId, tempName, tempBrand, tempPrice, tempSize, tempMaterial, tempGender, tempColor, tempSleeveType);

        // push back new data to container
        this->container.push_back(new_data);

        cout << "Data has been successfully created!\n";
    }

    void modifyData() {

        string target_id;
        cout << "Please select data to modify by ID\n";
        cin >> target_id;
        list<Shirt>::iterator target = searchByID(target_id);
        if (target == this->container.end()) {
            cout << "Data was not found!\n";
            return;
        }

        string tempName, tempBrand, tempSize, tempMaterial, tempGender, tempColor, tempSleeveType;
        double tempPrice;

        cout << "Data found. Please enter new data\n";
        cout << "Name : ";
        cin >> tempName;
        cout << "Brand : ";
        cin >> tempBrand;
        cout << "Price : ";
        cin >> tempPrice;
        cout << "Size : ";
        cin >> tempSize;
        cout << "Material : ";
        cin >> tempMaterial;
        cout << "Gender : ";
        cin >> tempGender;
        cout << "Color : ";
        cin >> tempColor;
        cout << "Sleeve Type : ";
        cin >> tempSleeveType;

        target->setName(tempName);
        target->setBrand(tempBrand);
        target->setPrice(tempPrice);
        target->setSize(tempSize);
        target->setMaterial(tempMaterial);
        target->setGender(tempGender);
        target->setColor(tempColor);
        target->setSleeveType(tempSleeveType);

        cout << "Data successfully modified!\n";

    }

    void deleteData() {

        string target_id;
        cout << "Please select data to modify by ID\n";
        cin >> target_id;

        list<Shirt>::iterator target = searchByID(target_id);
        char del_confirm_input;


        if (target == this->container.end()) {
            cout << "Data was not found!\n";
            return;
        }

        cout << "Data has been found:\n";

        // deletion confirmation
        cout << target->getId() << " | " << target->getName() << "\n\nConfirm to delete?(y/n)\n";
        cin >> del_confirm_input;
        if (del_confirm_input == 'y') {
            this->container.erase(target);
            cout << "Data has successfully been deleted\n";
        }
        else {
            cout << "Canceled data deletion\n";
        }
    }


    void showAll() {
        Table table;

        // Create data vector
        std::vector<std::vector<std::string>> data;

        // Header row
        data.push_back({"Product ID", "Name", "Brand", "Price", "Size", "Material", "Gender", "Color", "Sleeve Type"});

        // Iterate through container to populate data
        for (const auto& shirt : container) {
            data.push_back({
                shirt.getId(),
                shirt.getName(),
                shirt.getBrand(),
                std::to_string(shirt.getPrice()), // Convert price to string
                shirt.getSize(),
                shirt.getMaterial(),
                shirt.getGender(),
                shirt.getColor(),
                shirt.getSleeveType()
            });
        }

        // Call createRows method
        table.createRows(data.size(), data[0].size(), data);
    }

    // destructor
    ~DataController() {

    }

};
